recipe-client-addon/test/browser/browser_AddonStudies.js

```diff
@@ -282,3 +271,70 @@ compose_task(
     });
   }
 );
+
+compose_task(
+  AddonStudies.withStudies([
+    studyFactory({active: true, addonId: "does.not.exist@example.com", studyEndDate: null}),
+    studyFactory({active: true, addonId: "installed@example.com"}),
+    studyFactory({active: false, addonId: "already.gone@example.com", studyEndDate: new Date(2012, 1)}),
+  ]),
+  withInstalledWebExtension({id: "installed@example.com"}),
+  async function testInit([activeStudy, activeInstalledStudy, inactiveStudy]) {
+    await AddonStudies.init();
+
+    const newActiveStudy = await AddonStudies.get(activeStudy.recipeId);
+    ok(!newActiveStudy.active, "init marks studies as inactive if their add-on is not installed.");
+    ok(
+      newActiveStudy.studyEndDate,
+      "init sets the study end date if a study's add-on is not installed."
+    );
+
+    const newInactiveStudy = await AddonStudies.get(inactiveStudy.recipeId);
+    is(
+      newInactiveStudy.studyEndDate.getFullYear(),
+      2012,
+      "init does not modify inactive studies."
+    );
+
+    const newActiveInstalledStudy = await AddonStudies.get(activeInstalledStudy.recipeId);
+    Assert.deepEqual(
+      activeInstalledStudy,
+      newActiveInstalledStudy,
+      "init does not modify studies whose add-on is still installed."
+    );
+  }
+);
+
+compose_task(
+  AddonStudies.withStudies([
+    studyFactory({active: true, addonId: "installed@example.com", studyEndDate: null}),
+  ]),
+  withInstalledWebExtension({id: "installed@example.com"}),
+  async function testInit([study], [id, addonFile]) {
+    // Wait (or throw if we time out ) for the shield-study-ended event to
```

Comment:
You can just use `TestUtils.topicObserved` instead of rolling your own. I don't think the timeout is a good idea because it will likely intermittently fail on debug build runs, but if you want to keep using it, use `Promise.race()` with a different resolution value and then use that value once the result of `Promise.race` resolves.